docs/discovery-naming-conventions.md

```diff
@@ -0,0 +1,137 @@
+
+Backend systems other than Kubernetes (e.g. OpenStack) might support names that
+contain characters that are invalid according to the Kubernetes Service naming
+requirements. In these scenarios, discoverers will skip the service and inform
+the user via logs and metrics that the backend service name is not compatible
+with Gimbal.
+
+Even thought this might affect the user experience and cause friction, Gimbal
+prefers to be predictable and safe when it comes to managing backend services
+and endpoints, instead of using "magic" (replacing or removing characters, etc)
+to massage invalid names into valid ones.
+
+### Handling Discovered Names that are longer than 63 characters
+
+When the _Discovered Name_ is longer than 63 characters, it is shortened using
+the following process:
+
+1. Each component of the _Discovered Name_ gets the same number of maximum
+   characters allowed. The _Discovered Name_ has a total of three components,
+   and thus each component is allocated a maximum of 21 characters.
+
+2. Take the SHA256 hash of the _Discovered Name_ before shortening, and take the
+   first 6 characters of the resulting hash. This is called the _short hash_.
+
+3. Starting at the last _Discovered Name Component_, check whether it is longer
+   than the allocated number of characters. If it is, truncate the end of the
+   _Component_ and append `-${short-hash}`.
+
+4. If the resulting _Discovered Name_ is still longer than 63 characters, move
+   onto the next _Component_ and shorten using the _short hash_. Continue until
+   the resulting _Discovered Name_ is 63 characters or less.
+
+5. In the case that the _Discovered Name_ is longer than 63 characters after the
+   shortening process, the SHA256 hash produced in step 2 will be used as the
+   _Discovered Name_, prepended with `g-` to ensure the name starts with a
+   letter. This scenario will only happen if *all* _Components_ of the initial
+   _Discovered Name_ are longer than 21 characters, which is a rare case.
+
+#### Example
+
+- `${discoverer-prefix}`: `k8s`
+- `${service-name}`: `the-really-long-kube-service-name-that-is-exactly-63-characters`
+- `${backend-name}`: `us-east-cluster`
+
+1. The _Discovered Name_ has a total of three components. Thus, allocate 21 (61/3) characters to each.
+
+2. Take the SHA256 hash of `k8s-the-really-long-kube-service-name-that-is-exactly-63-characters-us-east-cluster`.
+
+    ```
+    SHA256 hash = cdbeb21c3261255a91ded1844754791ef1ec06ecbe5b2421ff2abe221950fc72
+    Short hash = cdbeb2
+    ```
+
+3. Check if the last _Component_ of the _Discovered Name_ goes over 21 characters:
+
+    ```
+    "us-east-cluster" has 15 characters. Move onto next component.
+    ```
+
+4. Check if the next _Component_ goes over 21 characters:
+
+    ```
+    "the-really-long-kube-service-name-that-is-exactly-63-characters" has 63 characters.
+    Shorten using the short hash by truncating and appending the short hash.
+    Result: "the-really-lon-cdbeb2".
```

Comment:
I think when appending the short hash, we shouldn't use a hyphen, they are nice to keep the main segments apart.

Reply:
Good call! I like this.

Reply:
Assuming we limit the prefix to <=3 characters, we should revamp this example to use 29 total characters